gov/politeia/proposals.go

```diff
@@ -91,373 +90,607 @@ func NewProposalsDB(politeiaURL, dbPath string) (*ProposalDB, error) {
-		return errDef
+// fetchRecordDetails fetches the record details of the given proposal tokens.
+func (db *ProposalsDB) fetchRecordDetails(tokens []string) (map[string]*recordsv1.Record, error) {
+	records := make(map[string]*recordsv1.Record, len(tokens))
+	for _, token := range tokens {
+		detailsReq := recordsv1.Details{
+			Token: token,
+		}
+		dr, err := db.client.RecordDetails(detailsReq)
+		if err != nil {
+			return nil, fmt.Errorf("Pi client RecordDetails err: %w", err)
+		}
+		records[token] = dr
 	}
 
-	defer atomic.StoreInt64(&db.lastSync, time.Now().UTC().Unix())
+	return records, nil
+}
 
-	// Retrieve and update all current proposals whose vote statuses is either
-	// NotAuthorized, Authorized and Started
-	numRecords, err := db.updateInProgressProposals()
-	if err != nil {
-		return err
+// fetchCommentsCounts fetches the comments counts for the given proposal tokens.
+func (db *ProposalsDB) fetchCommentsCounts(tokens []string) (map[string]uint32, error) {
+	commentsCounts := make(map[string]uint32, len(tokens))
+	paginatedTokens := paginateTokens(tokens, commentsv1.CountPageSize)
+	for i := 0; i < len(paginatedTokens); i++ {
+		cr, err := db.client.CommentCount(commentsv1.Count{
+			Tokens: paginatedTokens[i],
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Pi client CommentCount err: %w", err)
+		}
+		for token, count := range cr.Counts {
+			commentsCounts[token] = count
+		}
 	}
+	return commentsCounts, nil
+}
 
-	// Retrieve and update any new proposals created since the previous
-	// proposals were stored in the db.
-	lastProposal, err := db.lastSavedProposal()
-	if err != nil && err != storm.ErrNotFound {
-		return fmt.Errorf("lastSavedProposal failed: %v", err)
+// fetchTicketVoteSummaries fetches the vote summaries for the given proposal tokens.
+func (db *ProposalsDB) fetchTicketVoteSummaries(tokens []string) (map[string]ticketvotev1.Summary, error) {
+	voteSummaries := make(map[string]ticketvotev1.Summary, len(tokens))
+	paginatedTokens := paginateTokens(tokens, ticketvotev1.SummariesPageSize)
+	for i := 0; i < len(paginatedTokens); i++ {
+		sr, err := db.client.TicketVoteSummaries(ticketvotev1.Summaries{
+			Tokens: paginatedTokens[i],
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Pi client TicketVoteSummaries err: %w", err)
+		}
+		for token := range sr.Summaries {
+			voteSummaries[token] = sr.Summaries[token]
+		}
 	}
+	return voteSummaries, nil
+}
 
-	var queryParam string
-	if len(lastProposal) > 0 && lastProposal[0].TokenVal != "" {
-		queryParam = fmt.Sprintf("?before=%s", lastProposal[0].TokenVal)
+// paginateTokens paginates tokens in a matrix according to the provided
+// page size.
+func paginateTokens(tokens []string, pageSize uint32) [][]string {
+	n := len(tokens)/int(pageSize) + 1 // number of pages needed
+	ts := make([][]string, n)
+	page := 0
+	for i := 0; i < len(tokens); i++ {
+		if len(ts[page]) >= int(pageSize) {
+			page++
+		}
+		ts[page] = append(ts[page], tokens[i])
```

Comment:
Not handling edge cases well here.  Namely when `len(tokens)` is a multiple of `pageSize`.

```suggestion
// paginateTokens paginates tokens in a matrix according to the provided page
// size. pageSize must not be zero.
func paginateTokens(tokens []string, pageSize int) [][]string {
	n := len(tokens) / pageSize
	if len(tokens)%pageSize != 0 {
		n++
	}
	ts := make([][]string, n)
	var page int
	for i := range tokens {
		ts[page] = append(ts[page], tokens[i])
		if (i+1)%pageSize == 0 {
			page++
		}
```

Reply:
Hrmmmmm of course, sorry I missed this mate, my bad. Fixing this up and including the test cases

Reply:
No biggie.  Lots of loc always means saying bugs, nothing unexpected.

BTW, the test cases should probably include multiple page results.  The ones I suggested were limited